### PR TITLE
APP-474: [C] Different behavior and output of Notification Demo on various SDK clients

### PR DESCRIPTION
**Pull request by RostakaGmfun**
/cc @Slash32 @GamovCoder

### CODE REVIEW

File: notification/source/c/src/kaa_demo.c [hunk 1]
```diff
@@ -20,8 +20,10 @@
 #include <platform/kaa_client.h>
 #include <utilities/kaa_log.h>
 #include <kaa_notification_manager.h>
+#include <string.h>
 
 static kaa_client_t *kaa_client = NULL;
+kaa_list_t *topic_list = NULL;
```
Comment (Slash32) [inline notification/source/c/src/kaa_demo.c:26]: This seems to be better declared to be static.

**Review by Slash32 — APPROVED** [review]
Looks good, just a small nitpick

File: notification/source/c/src/kaa_demo.c (hunk 1 above)
Comment (RostakaGmfun) [reply notification/source/c/src/kaa_demo.c:26]: So, should I fix it?

File: notification/source/c/src/kaa_demo.c [hunk 2]
```diff
@@ -57,9 +59,28 @@ static int read_topic_id(void)
...
     return atoi(input);
 }
 
+static kaa_topic_t *find_topic(uint64_t topic_id)
+{
+    if (!topic_list || !kaa_list_get_size(topic_list)) {
+        return NULL;
+    }
+    kaa_list_node_t *it = kaa_list_begin(topic_list);
+    while (it) {
```
Comment (ohamov) [inline notification/source/c/src/kaa_demo.c:74]: `kaa_list` already has the necessary API for search:
 `kaa_list_node_t *kaa_list_find_next(kaa_list_node_t *from, match_predicate pred, void *context);`
Refer to `kaa_list.h`

File: notification/source/c/src/kaa_demo.c [hunk 3]
```diff
@@ -98,19 +126,21 @@ void on_topics_received(void *context, kaa_list_t *topics)
     demo_printf("Topic list was updated\r\n");
     show_topics(topics);
 
-    demo_printf("Type topic ID in order to subscribe on one:");
-    size_t topic_id = read_topic_id();
+    demo_printf("Enter topic id to subscribe to:\n");
+    demo_printf("Enter 'quit' to exit\n");
+    int topic_id = read_topic_id();
```
Comment (ohamov) [inline notification/source/c/src/kaa_demo.c:131]: topics ids are `int64_t` type.

File: notification/source/c/src/kaa_demo.c [hunk 4]
```diff
@@ -98,19 +126,21 @@ void on_topics_received(void *context, kaa_list_t *topics)
...
     while (it) {
-        kaa_topic_t *topic = (kaa_topic_t *) kaa_list_get_data(it);
+        kaa_topic_t *topic = (kaa_topic_t *)kaa_list_get_data(it);
         if (topic->subscription_type == OPTIONAL_SUBSCRIPTION && topic->id == topic_id) {
-            demo_printf("Subscribing to optional topic '%lu'\r\n", topic->id);
             err = kaa_subscribe_to_topic(kaa_client_get_context(client)->notification_manager, &topic->id, false);
             if (err) {
                 demo_printf("Failed to subscribe.\r\n");
+            } else {
+                demo_printf("Subscribed to optional topic '%lu'\r\n", topic->id);
```
Comment (ohamov) [inline notification/source/c/src/kaa_demo.c:143]: You can get UB. `s/%lu/%llu/g`

File: notification/source/c/src/kaa_demo.c [hunk 5]
```diff
@@ -71,6 +92,11 @@ static void on_notification(void *context, uint64_t *topic_id, kaa_notification_
     } else {
         demo_printf("Error: Received notification's body is null\r\n");
     }
+
+    kaa_topic_t *topic = find_topic(*topic_id);
```
Comment (ohamov) [inline notification/source/c/src/kaa_demo.c:96]: Use existing API (see the comment above)

File: notification/source/c/src/kaa_demo.c [hunk 6]
```diff
@@ -57,9 +59,28 @@ static int read_topic_id(void)
     }
 
     input[input_length] = '\0';
+    if (strcmp(input, "quit") == 0) {
+        exit(EXIT_SUCCESS);
+    }
     return atoi(input);
```
Comment (ohamov) [inline notification/source/c/src/kaa_demo.c:65]: topic id type is not `int`

File: notification/source/c/src/kaa_demo.c (hunk 1 above)
Comment (Slash32) [reply notification/source/c/src/kaa_demo.c:26]: @RostakaGmfun, yes, please fix this and other comments to the PR within a new PR.

File: notification/source/c/src/kaa_demo.c (hunk 4 above)
Comment (RostakaGmfun) [reply notification/source/c/src/kaa_demo.c:143]: I suppose the safest way to print topic ids is by using PRid64

**Pull request merged**